test/suites/container_devices_nic_routed.sh

```diff
@@ -66,6 +66,24 @@ test_container_devices_nic_routed() {
     mtu=1600
   lxc start "${ctName}"
 
+  # Container has started.
+  # It should no longer be possible to start another container which has routed nic devices with the same IP addresses.
+  lxc init testimage dummy
+  lxc config device add dummy dummyeth4 nic \
+      name=dummyeth4 \
+      nictype=routed \
+      parent=${ctName} \
+      ipv4.address="192.0.2.1${ipRand}"
+  ! lxc start dummy
+  lxc config device remove dummy dummyeth4
+  lxc config device add dummy dummyeth6 nic \
+      name=dummyeth6 \
+      nictype=routed \
+      parent=${ctName} \
+      ipv6.address="2001:db8::1${ipRand}"
+  ! lxc start dummy
+  lxc delete dummy -f
+
```

Comment:
@stgraber This test is passing because the dummy container fails to start. However, it is currently failing for the wrong reason. It is failing with:
```
Error: Failed to start device "dummyeth4": Failed adding host route "192.0.2.14/32": Failed to run: ip -4 route add table main 192.0.2.14/32 dev veth14a2a1b9: RTNETLINK answers: File exists
```
but it should be:
```
Error: Failed to start device "dummyeth4": ipv4.address "192.0.2.17" is already in use
```

It works fine when running it manually so I think I'm missing something about how the test environment is set up. Just not sure what that is!

Reply:
The address you're setting here is already in use by the previous container.

Reply:
> The address you're setting here is already in use by the previous container.

Yes that's what I'm testing. The `lxc start dummy` command should fail at device start up because the test container `${ctName}` is already using it.

> As your test is self-contained, I think it should be moved prior to the other test (so at line 55 or so) so it doesn't end up conflicting with it.

This test is meant to show that the dummy container cannot start because it has a device attached that is configured to use an IP address that is already in use by another container. I can move to line 55 but will have to start a separate dummy container in order to occupy the addresses I'm testing against.

Reply:
Ah right, that makes sense. Could you add a `ip -4 route show` to your test? That should let us see what the routing table is like. In general I would have expected that `ip route add` to succeed unless it's somehow adding the address to the same veth.

Reply:
Ah, so the issue may be from both of those using the same `parent`. When `parent` is set, LXD will add routing entries to that device so it can respond to the address. Multiple instances sharing the same `parent` will likely cause that early failure.

As `parent` isn't strictly required, maybe your test will yield the correct result if your new container doesn't have the `parent` property set.

Reply:
This change skips the ARP/NDP check if the parent is not set (per https://github.com/lxc/lxd/pull/9758#discussion_r780197680 - see https://github.com/lxc/lxd/pull/9758/files#diff-e5d2c9ea3e0d86c4f244b314acd4940e9fabfeb2dc16c0a457c849978a317a94R308). So removing the parent would skip what we're trying to test unfortunately.

Reply:
`ip -4 route show` has:
```
+ ip -4 route show
default via 192.168.1.1 dev wlp0s20f3 proto static metric 600 
10.127.151.0/24 dev lxdbr0 proto kernel scope link src 10.127.151.1 
10.127.151.129 dev vethe0745975 scope link 
169.254.0.0/16 dev wlp0s20f3 scope link metric 1000 
192.0.2.16 dev veth81b94d2e scope link 
192.0.3.0/24 via 192.0.2.16 dev veth81b94d2e 
192.168.1.0/24 dev wlp0s20f3 proto kernel scope link src 192.168.1.100 metric 600 
```

Reply:
Been playing with this branch a bunch, the real issue here is that you do have an IP conflict but your test function isn't noticing it, which then causes us to keep moving and hit the route conflict issue.

Reply:
Right, so that test just can't work to detect a conflict on the same machine, though the route add later on does catch it.

Basically what's going on is that your arping and ndp is headed out through the `ntXYZ` interface which is meant to be the external interface of the system, but the existing instance isn't reached through that, it's reached through its `vethXYZ` interface instead. Now if you had a conflicting instance on a different machine that you normally would reach through `ntXYZ`, the test should work fine, but there's no way this check will work to find a conflict on the local machine.

As a result, I think we should just axe the test entirely on this one.

@tomponline thoughts if you happen to be lurking.

Reply:
Could you use a temporary dummy or bridge as the test patent interface?

Reply:
I did try this with a bridge network but got the same RTNETLINK error:

```
  lxc network create lxdbr1
  sysctl net.ipv6.conf.lxdbr1.proxy_ndp=1

  lxc init testimage c1
  lxc config device add c1 n1 nic \
    nictype=routed \
    parent=lxdbr1 \
    ipv4.address="192.0.2.1${ipRand}" \
    ipv6.address="2001:db8::1${ipRand}"
  lxc start c1

  lxc init testimage c2
  lxc config device add c2 n1 nic \
    nictype=routed \
    parent=lxdbr1 \
    ipv4.address="192.0.2.1${ipRand}"
  ! lxc start c2

  lxc config device remove c2 n1
  lxc config device add c2 n1 nic \
    nictype=routed \
    parent=lxdbr1 \
    ipv6.address="2001:db8::1${ipRand}"
  ! lxc start c2

  lxc delete c1 -f
  lxc delete c2
  lxc network delete lxdbr1
```

Reply:
You will need to add the IP to the bridge interface for duplicate detection outside of lxds containers. Otherwise ull get a duplicate route error.

Reply:
I don't think that can work unless we start using nested containers or a different namespace.

Basically this check can't work on a system which has a static route for the address already somewhere.
It should work perfectly fine if the address is in use on another system though as that will normally reach the external interface and broadcast ARP/NDP there with the host holding the instance responding to that.

We could probably test this still but it'd get a bit weird with:
 - Create a bridge device with a test /24 subnet on it
 - Connect a container using nictype=bridged to that bridge
 - Add the test IPv4 address inside of that test container
 - Now create another container using nictype=routed with the parent= set to the same bridge device as the existing container

This now gives us a "host" which doesn't have any static routes for the routed IP address and the ARP/NDP test should broadcast through the bridge, get a hit and refuse to reuse the address.

Reply:
Sounds good!

Reply:
You could also use another container with a p2p nic and static nic host_name setting with a manually configured IP inside it and use the static host_name interface as the parent of the routed nic instance.

Reply:
I've removed the tests (see https://github.com/lxc/lxd/pull/9758#issuecomment-1011700642).